### DISCUSSION TITLE
Help with CephFS mirroring (docs are incomplete and confusing)

### DISCUSSION THREAD

**Discussion by dhess** [rook/rook#13351] [Q&A]
Hi, we're trying to create a one-way mirror of CephFS filesystem from one Rook cluster (the "primary" cluster) to another (the "secondary" cluster). The Rook clusters are running in different Kubernetes clusters.

I'm finding the [Rook filesystem mirroring documentation](https://rook.io/docs/rook/latest-release/Storage-Configuration/Shared-Filesystem-CephFS/filesystem-mirroring/) to be pretty confusing. Enabling mirroring on the primary cluster filesystem, starting the mirror daemon (via the `CephFilesystemMirror` CRD), and creating the primary cluster bootstrap token are all straightforward, but from there, I can't figure out what the correct next steps are:

1. Presumably the source filesystem (let's call it `cephfs-0`) needs to be defined in the secondary cluster. Is it necessary to set `mirroring: enabled: true` on the secondary cluster filesystem, as well? As we only want a one-way mirror, where the primary syncs to the secondary but not vice versa, I assume the answer is "no," at least in our specific case.
2. Does the `CephFilesystemMirror` CRD need to be created in both the primary and the secondary clusters, or only in the primary? By my reading, the Rook docs say that it only needs to be run in the primary, but the secondary cluster steps aren't very detailed and refer to the Ceph docs, which don't know or say anything about Rook, of course, so I'm not 100% certain about this.
3. By my reading, the Rook docs say that you need to create a bootstrap token in the primary cluster, then [import the primary's bootstrap token into the secondary cluster](https://rook.io/docs/rook/latest-release/Storage-Configuration/Shared-Filesystem-CephFS/filesystem-mirroring/#import-the-token-in-the-destination-cluster). However, when I tried this on my secondary cluster (after creating the secondary cluster remote mirror user via the `ceph fs authorize cephfs-0 client.mirror_remote / rwps` command from the `rook-ceph-tools` pod CLI, and also loading the mirroring module and enabling mirroring on the target filesystem, both of which Ceph complained about), the import command just hung forever. Upon a [closer reading of the Ceph boostrap peers docs](https://docs.ceph.com/en/latest/dev/cephfs-mirroring/#bootstrap-peers), it sounds like this is backwards, anyways: ref "*Adding a peer (via peer_add) requires the peer cluster configuration and user keyring to be available in the primary cluster (manager host and hosts running the mirror daemon). This can be avoided by bootstrapping and importing a peer token.*" and "*Import the bootstrap token in the primary cluster via ...*" from the Ceph documentation. Importing the secondary token into the primary seems to be the procedure followed by @ahmedTouati in this somewhat helpful related issue here: https://github.com/rook/rook/issues/9151
4. Once one successfully imports the peer bootstrap token into one cluster or the other (or both), how does one specify the address of the peer cluster? In our case, it should be easy enough to set up a load balancer or ingress for the services that need to be exposed to the peer cluster, but it's not at all clear, even from the Ceph documentation, how one specifies the remote cluster address when using the peer bootstrapping process. (When using the `ceph fs snapshot mirror peer_add` command rather than the bootstrap token command, you can specify the `remote_mon_host` address, but in the case of the bootstrap token command, it seems you just provide a human-readable string, which isn't treated as an IP address or DNS name.)
5. Speaking of load balancers and/or ingress, which services, exactly, need to be exposed to the peer cluster? Both the `mgr` and `mon` services, or just one of them?

**Comment by travisn** [comment]
@gauravsitlani Since you wrote the cephfs mirroring docs, could you take a look?

**Reply by gauravsitlani** [reply]
hey @travisn @dhess sure, I'll take a look and get back as soon as possible

**Comment by dhess** [comment]
Judging by this thread on the Ceph mailing list:

https://lists.ceph.io/hyperkitty/list/ceph-users@ceph.io/message/UWQLK6YLVBIYTA6KLLDUYN3ZXNDOLGBZ/

it seems plausible that the answer to question 4 ("how does one specify the address of the peer cluster?") is that the peer token includes the mon IPs from its cluster's `/etc/ceph/ceph.conf`, so that's how the primary cluster will try to connect to the secondary. If that's true, it would explain why the import command on the primary hangs, because I don't have connectivity to the secondary's mon IPs from the primary, as I'm running Cilium BGP for pod and load balancer routing. I will try to fix that now.

**Reply by gauravsitlani** [reply]
@dhess yes you'll need to have proper connectivity between both the clusters.

**Reply by dhess** [reply]
@gauravsitlani Thanks. I'm not planning to implement CephFS mirroring via Ceph/Rook anymore, so there's no urgency on my end, but I do think the Rook docs need some clean up for this feature.

**Comment by deborggraever** [comment]
We are running more or less into the same issue / questions.

We have two kubernetes clusters which run rook/ceph with CephFS.
We want to setup mirroring between both clusters (active/passive)

Documentation is indeed very unclear, so after lot of searching we got somewhere, but we got stuck.
It seems the the bootstrap token use the service ip's instead of host ip's.
We also run multus, so the monitor pods got two network interfaces but seem to have bound the service the the internal pod network only.

**Comment by akash123-eng** [comment]
We are facing the same issue.

after following all steps at the end ceph fs snapshot mirror peer botstrap import command hangs forever.
there are no steps/prequistes given on how networking should be setup in order to mirroring to work i.e which ports should be open between both rook-ceph clusters etc.
@travisn  please help here. many users are facing issue with cephfs mirroring

**Reply by travisn** [reply]
The challenge is that the contributor who created the cephfs mirroring docs is no longer with the project. It is a very advanced feature and current maintainers may not be able to help with the details. We would look forward to updating the docs as you or other community members find fixes. Hopefully you can find help from the upstream ceph community as well.

I've heard of some using Volsync with CephFS, you might also consider that if the cephfs mirroring remains elusive.

**Reply by deborggraever** [reply]
I'm not sure if the issue is only the documentation.
I went trough the code to try find a solution for our problem.

We are using multus, which might also be cause of the issue here. We did this because we want to split the traffic between the ceph cluster and the public k8s traffic.

So tracing back the issue here:
We unbase64 the bootstrap token and notice that the peer ip's are the k8s monitoring pod service ip's, so not reachable from another cluster.
We changed these, base64 again and try to import on the other side. This did not work.

Next we try to see if the other side is reachable.
Ping OK, curl on the monitoring ports give connection refused.

Next we try to see if the pod works on the correct ip's

Doing "ip a" we see we have a k8s pod IP + the ceph cluster IP, so this part is ok.
We ping the monitoring pod ceph cluster ip on other side, ok.
We curl the monitor ports on the pod ip, also working
We curl the monitor ports on the ceph cluster ip, connection refused.

The /etc/ceph config files are empty. So somehow we need to tell the monitor pods to listen on both addresses.
The --public-addr is the service address cause we don't initialized using hostnetwork. If u understand the comments in the code correctly, we should not change this when the cluster is already running.

So to fix the issue we need to get the monitor pods to listen on both the pod ip & ceph cluster ip.

**Reply by BlaineEXE** [reply]
Monitors cannot listen on multiple addresses. This is a Ceph requirement. Mons must also have a static IP, and Rook's only method so far of accomplishing that has been to use Services for each monitor. Because monitors are not in the data path, this still allows multus to speed up communications.

But it does mean that it is much more complicated to set up inter-cluster communication where multus is involved. Network routing needs to be planned and configured carefully to allow for communication, and we don't have documents for doing this, as I don't know of anyone who has done it.

You may want to consider switching your cluster to host networking to help ease the network complexity you will face.

**Reply by deborggraever** [reply]
@BlaineEXE we spend all day yesterday breaking down the clusters and rebuilding them.
Currently we use host network instead of multus.
The monitor pods now have 3 ip addresses from the node, the pod network, the ceph public network and ceph cluster network.
We also added the rook-config-override, so the /etc/ceph/ceph.conf with network config is present.

Still the ceph mon is running with command line arg "--public-addr" which belongs to the pod network. So from the other cluster we can only access the monitor for replication over the pod network instead of the ceph network.

Any advice here?

We also tried the "addressRanges" with both multus and host network, but it seems to make no difference what we put there.
```
 network:
    provider: host
    addressRanges:
      public:
        - "192.168.100.0/24"
        - "192.168.101.0/24"
      cluster:
        - "192.168.200.0/24"
```

**Reply by travisn** [reply]
If the mons are configured with host network, the `--public-addr` would be the host IP. Did you change the host network setting in the running cluster? If so, the mons will retain their pod IPs and you would need to [failover the mons](https://rook.io/docs/rook/latest/Storage-Configuration/Advanced/ceph-mon-health/#failing-over-a-monitor) to get mons based on the host IP.

**Reply by BlaineEXE** [reply]
I believe Travis' info above is right and will help. 

Additionally, more context: if you look at the "Known limitations with Multus" section of the doc, you will see this note:

> Daemons leveraging Kubernetes service IPs (Monitors, Managers, Rados Gateways) are not listening on
the NAD specified in the `selectors`. Instead the daemon listens on the default network, however the
NAD is attached to the container, allowing the daemon to communicate with the rest of the cluster.

This also means that monitors will not be affected by changes to the `addressRanges` configs that change the cluster `public_network` and `cluster_network` configs. 

For host networking, I do recommend that you set the `addressRanges.public` and `addressRanges.cluster` values to match the host networks you want the Ceph cluster to use, however.

**Comment by cra1gg** [comment]
Having the same issue. Tried both what the rook docs said as well as importing the bootstrap token on the source cluster and neither are showing up as peers of eachother
